lib/agent/socket_util.js

```diff
@@ -110,13 +111,22 @@ exports.secureSocket = function (socket, host, agent, mock)
     Logger.getInstance().trace('socket reused = %s', socket.isSessionReused());
   };
 
-  // when the socket is secure, perform additional validation
-  socket.on('secure', validate);
+  if (socket.authorized) { // when socket is authorized we can check it now
+    socket.cork();
+    validate();
+    return {
+      socket,
+      socketError,
+    };
+  } else {
+    // when the socket is secure, perform additional validation
+    socket.on('secure', validate);
```

Comment:
What is the difference between this line and 116?

Reply:
first if branch performs synchronized validation and else performs validation asynchronously